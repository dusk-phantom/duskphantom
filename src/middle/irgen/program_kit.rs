// Copyright 2024 Duskphantom Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// SPDX-License-Identifier: Apache-2.0

use anyhow::Result;

use crate::middle::ir::FunPtr;
use crate::middle::irgen::value::Value;
use crate::utils::frame_map::FrameMap;
use crate::{frontend, middle};

/// Kit for translating a program to middle IR
pub struct ProgramKit<'a> {
    pub env: FrameMap<'a, String, Value>,
    pub fun_env: FrameMap<'a, String, FunPtr>,
    pub program: &'a mut middle::Program,
}

/// A program kit (top level) can generate declarations
impl<'a> ProgramKit<'a> {
    pub fn gen(mut self, program: &frontend::Program) -> Result<()> {
        self.gen_library_function();
        for decl in program.module.iter() {
            self.gen_global_decl(decl)?;
        }
        for decl in program.module.iter() {
            self.gen_impl(decl)?;
        }
        Ok(())
    }
}
