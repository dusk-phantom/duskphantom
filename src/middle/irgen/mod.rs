use std::collections::HashMap;

use crate::{frontend, middle};
use anyhow::Result;
use program_kit::ProgramKit;

mod constant;
mod function_kit;
mod gen_binary;
mod gen_const_binary;
mod gen_const_expr;
mod gen_const_unary;
mod gen_expr;
mod gen_global_decl;
mod gen_impl;
mod gen_inner_decl;
mod gen_library_function;
mod gen_stmt;
mod gen_unary;
mod program_kit;
mod value;
mod value_type;

/// Generate middle IR from a frontend AST
pub fn gen(program: &frontend::Program) -> Result<middle::Program> {
    let mut result = middle::Program::new();
    ProgramKit {
        program: &mut result,
        env: &mut vec![HashMap::new()],
        fun_env: &mut vec![HashMap::new()],
    }
    .gen(program)?;
    Ok(result)
}
