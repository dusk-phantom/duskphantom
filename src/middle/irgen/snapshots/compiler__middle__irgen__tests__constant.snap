---
source: src/middle/irgen/mod.rs
expression: llvm_ir
---
@PI = dso_local constant float 3.1415925
define i32 @main() {
%entry:
%alloca_2 = alloca i32
%load_5 = load float, ptr @PI
%fptoi_6 = fptosi float %load_5 to i32
store i32 %fptoi_6, ptr %alloca_2
br label %exit

%exit:
%load_3 = load i32, ptr %alloca_2
ret i32 %load_3


}
